File: lib/utils/s3PathDefinitions/dataStorePaths.dart
///This class is only for a common definition of paths between the apps.
///
///It has to be synchronized in the application's lib folder to be used.

import 'dart:core';

/// -- important --
///todo: when integrating into the app the method has to be tested
///this was not possible when creating the cluss due to a missing runnable environment

///enum of available paths to avoid errors due to wrong string spelling
enum DataStorePaths {
  userPicPath,
  levelPicPath,
  levelCustomDataPicPath,
  interventionPicPath,
  docPdfPath,
  docPicPath,
  interventionSurveyPicPath,
  questionPicPath,
  questionOptionPicPath,
  questionPicAnswerPath,
  questionAudioAnswerPath,
  appliedInterventionPicPath,
  entityPicPath,
  taskPicPath,
  taskAudioPath
}

///Map of ontoligies including the replacable parameters
final Map<DataStorePaths, dynamic> databaseOntologies = {
  DataStorePaths.userPicPath: {
    "path": "userFiles/userID/pic.png",
    "toBeReplaced": ["userID"]
  },
  DataStorePaths.levelPicPath: {
    "path": "levelFiles/levelID/pic.png",
    "toBeReplaced": ["levelID"],
  },
  DataStorePaths.levelCustomDataPicPath: {
    "path": "levelFiles/levelID/customDataFiles/customDataID/pic.png",
    "toBeReplaced": ["levelID", "customDataID"],
  },
  DataStorePaths.interventionPicPath: {
    "path": "interventionFiles/interventionID/pic.png",
    "toBeReplaced": ["interventionID"],
  },
  DataStorePaths.docPdfPath: {
    "path": "documentFiles/documentID/pdf.pdf",
    "toBeReplaced": ["documentID"],
  },
  DataStorePaths.docPicPath: {
    "path": "documentFiles/documentID/pic.png",
    "toBeReplaced": ["documentID"],
  },
  DataStorePaths.interventionSurveyPicPath: {
    "path": "interventionFiles/interventionID/surveyFiles/surveyID/pic.png",
    "toBeReplaced": ["interventionID", "surveyID"],
  },
  DataStorePaths.questionPicPath: {
    "path":
        "interventionFiles/interventionID/surveyFiles/surveyID/questionFiles/questionID/pic.png",
    "toBeReplaced": ["interventionID", "surveyID", "questionID"],
  },
  DataStorePaths.questionOptionPicPath: {
    "path":
        "interventionFiles/interventionID/surveyFiles/surveyID/questionFiles/questionID/optionFiles/optionID/pic.png",
    "toBeReplaced": ["interventionID", "surveyID", "questionID", "optionID"],
  },
  DataStorePaths.questionPicAnswerPath: {
    "path":
        "appliedInterventionFiles/appliedInterventionID/executedSurveyFiles/executedSurveyID/questionFiles/questionID/pic.png",
    "toBeReplaced": ["appliedInterventionID", "executedSurveyID", "questionID"],
  },
  DataStorePaths.questionAudioAnswerPath: {
    "path":
        "appliedInterventionFiles/appliedInterventionID/executedSurveyFiles/executedSurveyID/questionFiles/questionID/audio.mp3",
    "toBeReplaced": ["appliedInterventionID", "executedSurveyID", "questionID"],
  },
  DataStorePaths.appliedInterventionPicPath: {
    "path": "appliedInterventionFiles/appliedInterventionID/pic.png",
    "toBeReplaced": ["appliedInterventionID"],
  },
  DataStorePaths.entityPicPath: {
    "path": "entityFiles/entityID/pic.png",
    "toBeReplaced": ["entityID"],
  },
  DataStorePaths.taskPicPath: {
    "path": "taskFiles/taskID/picID.png",
    "toBeReplaced": ["taskID", "picID"]
  },
  DataStorePaths.taskAudioPath: {
    "path": "taskFiles/taskID/audioID.mpr",
    "toBeReplaced": ["taskID", "audioID"],
  },
};

///method passed the path name and the required paramters
///throws an error if parameters not fitting the paths are passed
String dataStorePath(DataStorePaths path, List<String> parameters) {
  String rawPath = databaseOntologies[path]["path"];
  List<String> replacementParameters = databaseOntologies[path]["toBeReplaced"];

  String filledPath = rawPath;

  if (parameters.length == replacementParameters.length) {
    for (int i = 0; i < parameters.length; i++) {
      filledPath =
          filledPath.replaceAll(replacementParameters[i], parameters[i]);
    }
    return filledPath;
  } else {
    throw new FormatException(
        "Wrong parameters passed for dataStorePath $path");
  }
}

File: analytics-app/backend/queries/data_store_paths.py


question_answer_pic_path = {
    "path":
        "organization/organizationID/appliedInterventionFiles/appliedInterventionID/executedSurveyFiles/executedSurveyID/questionFiles/questionID/pic.png",
    "toBeReplaced": ["organizationID", "appliedInterventionID", "executedSurveyID", "questionID"],
}

question_answer_audio_path = {
    "path":
        "organization/organizationID/appliedInterventionFiles/appliedInterventionID/executedSurveyFiles/executedSurveyID/questionFiles/questionID/audio.aac",
    "toBeReplaced": ["organizationID", "appliedInterventionID", "executedSurveyID", "questionID"],
}

#method to get the path for question_answer_audio_path
def get_question_answer_audio_path(organizationID, appliedInterventionID, executedSurveyID, questionID):
    path = question_answer_audio_path["path"]
    path = path.replace(question_answer_audio_path["toBeReplaced"][0], organizationID)
    path = path.replace(question_answer_audio_path["toBeReplaced"][1], appliedInterventionID)
    path = path.replace(question_answer_audio_path["toBeReplaced"][2], executedSurveyID)
    path = path.replace(question_answer_audio_path["toBeReplaced"][3], questionID)
    return path

#method to get the path for question_answer_pic_path
def get_question_answer_pic_path(organizationID, appliedInterventionID, executedSurveyID, questionID):
    path = question_answer_pic_path["path"]
    path = path.replace(question_answer_pic_path["toBeReplaced"][0], organizationID)
    path = path.replace(question_answer_pic_path["toBeReplaced"][1], appliedInterventionID)
    path = path.replace(question_answer_pic_path["toBeReplaced"][2], executedSurveyID)
    path = path.replace(question_answer_pic_path["toBeReplaced"][3], questionID)
    return path